metaseq/trainer.py

```diff
@@ -966,6 +977,31 @@ def clip_grad_norm(
             skip_gradient_update_on_clip_norm=skip_gradient_update_on_clip_norm,
         )
 
+    def skip_spike(self, logging_outputs, ewm_ratio_to_skip_batch):
+        if ewm_ratio_to_skip_batch == -1:
+            return
+        loss_sum = sum(log.get("loss", 0) for log in logging_outputs)
+        sample_size = sum(log.get("sample_size", 0) for log in logging_outputs)
+        loss_t = float(loss_sum / sample_size / math.log(2))
+
+        if self._ewm_loss is None:
+            self._ewm_loss = loss_t
+
+        ewm_t_1 = self._ewm_loss
+        ewm_t = ewm(loss_t, ewm_t_1, span=9)
+        ewm_ratio = loss_t / ewm_t
+
+        if ewm_ratio > ewm_ratio_to_skip_batch:
+            raise SpikeError(
+                f"Skip batch as we encountered a loss spike. In "
+                f"num_update: {self.get_num_updates()} the loss is {loss_t:.2f}. "
+                f"The ewm for the loss was only at {ewm_t:.2f} . "
+                f"The loss to ewm loss ratio is {ewm_ratio:.2f}, which is higher than "
+                f"ewm_ratio_to_skip_batch of {ewm_ratio_to_skip_batch} ."
+            )
+
+        self._ewm_loss = ewm_t
```

Comment:
Note: We are only updating the ewm if we keep the update. Otherwise the ewm would adapt quickly to bad batches and then stop skipping them.

Reply:
Let's leave a comment explaining this please